common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -194,122 +215,143 @@ private IRequestTelemetryCache createLastRequestTelemetryCache(@NonNull final Co
-     */
-    public void flush(final String correlationId, final BaseException baseException) {
-        flush(correlationId, baseException == null ? null : baseException.getErrorCode());
-    }
+    public void flush(@NonNull final BaseCommand command, @NonNull final CommandResult commandResult) {
+        final String methodName = ":flush";
 
-    /**
-     * Flush the telemetry data associated to the correlation id to the {@link android.content.SharedPreferences} using the {@link SharedPreferencesLastRequestTelemetryCache}.
-     * Removes the telemetry associated to the correlation id from the telemetry map,
-     * and saves it to the cache (SharedPreferences) as the last request telemetry.
-     *
-     * @param correlationId      correlation id of the request
-     * @param acquireTokenResult the result obtained from the acquire token call
-     */
-    public void flush(final String correlationId, final AcquireTokenResult acquireTokenResult) {
-        final String errorCode = TelemetryUtils.errorFromAcquireTokenResult(acquireTokenResult);
-        flush(correlationId, errorCode);
-    }
+        final String correlationId = command.getParameters().getCorrelationId();
 
-    /**
-     * Flush the telemetry data associated to the correlation id to the {@link android.content.SharedPreferences} using the {@link SharedPreferencesLastRequestTelemetryCache}.
-     * Removes the telemetry associated to the correlation id from the telemetry map,
-     * and saves it to the cache (SharedPreferences) as the last request telemetry.
-     *
-     * @param correlationId correlation id of the request
-     * @param errorCode     error that may have occurred during the request
-     */
-    public void flush(final String correlationId, final String errorCode) {
-        final String methodName = ":flush";
         if (mTelemetryMap == null || correlationId == null) {
             return;
         }
 
-        RequestTelemetry currentTelemetry = mTelemetryMap.get(correlationId);
+        CurrentRequestTelemetry currentTelemetry = mTelemetryMap.get(correlationId);
+
         if (currentTelemetry == null) {
             return;
         }
 
-        RequestTelemetry lastTelemetry = setupLastFromCurrent(currentTelemetry);
-        lastTelemetry.putTelemetry(Schema.Key.CORRELATION_ID, correlationId);
-        lastTelemetry.putTelemetry(Schema.Key.ERROR_CODE, errorCode);
+        // load the last request object from cache
+        LastRequestTelemetry lastRequestTelemetry = loadLastRequestTelemetryFromCache();
+
+        // We did not have a last request object in cache, let's create a new one and derive
+        // fields from current request where applicable
+        if (lastRequestTelemetry == null) {
+            lastRequestTelemetry = new LastRequestTelemetry(currentTelemetry.getSchemaVersion());
+            lastRequestTelemetry = (LastRequestTelemetry) lastRequestTelemetry.derive(currentTelemetry);
```

Comment:
Why are we duplicating the data?

Reply:
duplicating?

Reply:
The data there needs to be on both current request and last request telemetry objects. `derive` is just a helper method to grab all the common platform fields easily without explicitly calling emit for each one.

Reply:
It was sort of confusing to me to have the same telemetry object in both fields (current and last).

I would expect us to only send 'previous failing requests' in this 'Last request' fields. Isn't that the case?

(If this particular request is not sent and it failed, only then we added it to the cache, and the next request would pick it up and send it to the server).

Reply:
Yes, that's right, we only send previous failing requests/errors. However, we also have to abide by the schema and send all fields/delimiters regardless of whether they have value or not. So it there are no last request telemetry available in the cache, then we need to send something like this:

`2|0|||`

i.e. We still send the schema version all delimiters